Add getProvider to InjectedConnector options

ekenigs opened the pull request:
## Description

Add `getProvider` to `InjectedConnector` options. This is necessary for many EVM (MetaMask) compatible wallets in the Polkadot ecosystem that are injecting a provider not to `window.ethereum`. 

- [SubWallet](https://chrome.google.com/webstore/detail/subwallet-polkadot-extens/onhogfjeacnfoofkfgppdlbmlmnplgbn) : `window.SubWallet`
- [Talisman](https://chrome.google.com/webstore/detail/talisman-polkadot-wallet/fijngjgcjhjmmpcmkeiomlglpeiijkld): `window.talismanEth`
- [Enkrypt](https://chrome.google.com/webstore/detail/enkrypt-ethereum-polkadot/kkpllkodjeloidieedojogacfhpaihoh): `window.enkrypt.providers.ethereum`

## Additional Information

- [x] I read the [contributing docs](/wagmi-dev/wagmi/blob/main/.github/CONTRIBUTING.md) (if this is your first contribution)

jxom commented on .changeset/pink-pillows-dream.md:
```diff
@@ -0,0 +1,5 @@
+'@wagmi/core': patch
+---
+
+Added getProvider to InjectedConnector options
```
Nice changeset! Maybe we can just add some quick usage as well.

```suggestion
Added `getProvider` to `InjectedConnector` options.

Example: 

```ts
import { InjectedConnector } from '@wagmi/core/connectors/injected'

const connector = new InjectedConnector({
  options: {
    name: 'SubWallet',
    getProvider: () =>
      typeof window !== 'undefined' ? window.SubWallet : undefined,
  },
})
```

jxom commented on packages/core/src/connectors/injected.ts:
```diff
@@ -39,7 +41,7 @@ export class InjectedConnector extends Connector<
   readonly id: string
   readonly name: string
-  readonly ready = typeof window != 'undefined' && !!window.ethereum
+  readonly ready: boolean = false
```
Probably don't need a default now since we assign in the constructor.

```suggestion
  readonly ready: boolean
```

ekenigs commented:
@jxom thanks for your review.

tmm reviewed (commented):
Added a comment. Would also be great if we could pull some of the behavior around the `MetaMaskConnector` into this so it could be completely powered by `getProvider` as well (let me know if you want to chat about this more).

ekenigs commented:
@tmm thanks for your review and suggestions. As you suggested I have removed `getProviderFromOptionsOrWindow` and added the default `getProvider` to the options. But I needed to add an extra interface to make `getProvider` not optional in `this.options`.

Also using `getProvider` in the `MetaMaskConnector` connector. 

Please check if it's what you like.

tmm commented:
Created new PR #1406 to continue work since I wasn't able to make updates to this one.
